tests/virt_autotest/xen_guest_irqbalance.pm

```diff
@@ -0,0 +1,192 @@
+        }
+
+        # Get the network interrupts distribution over CPUs in guest prior to network loads
+        my @initial_total_irqs_on_cpu = get_nic_irqs_distribution_from_guest($guest, $nproc);
+
+        # Generate network load in guest
+        assert_script_run("ssh root\@$guest \"ping -f -c 500 download.opensuse.org\"");
+
+        # re-caculate the network interrupts distribution
+        # the increased NIC IRQs should be distributed in CPU cores in balance
+        my @total_irqs_on_cpu_after_network_download = get_nic_irqs_distribution_from_guest($guest, $nproc);
+        my @increased_irqs_on_cpu;
+        for (my $cpu_id = 0; $cpu_id < $nproc; $cpu_id++) {
+            $increased_irqs_on_cpu[$cpu_id] = $total_irqs_on_cpu_after_network_download[$cpu_id] - $initial_total_irqs_on_cpu[$cpu_id];
+            die("The value of network interrupts from download for CPU" . $cpu_id . " is 0") if $increased_irqs_on_cpu[$cpu_id] == 0;
+        }
+        record_info("IRQs of downloads distribution", "@increased_irqs_on_cpu");
+    }
+
+    restore_original_guests();
+
+}
+
+#set up guest test environment to run irqbalance test
+sub prepare_guest_for_irqbalance {
+    my $vm_name = shift;
+
+    #4 or more vcpu is needed
+    my $nproc = script_output "virsh vcpucount $vm_name | grep current | grep config | awk '{ print \$3 }'";
+    if ($nproc < 4) {
+        assert_script_run "virsh dumpxml --inactive $vm_name > $vm_name.xml";
+        script_run "virsh destroy $vm_name";
+        assert_script_run "xmlstarlet edit -L -d /domain/vcpu $vm_name.xml";
+        assert_script_run "xmlstarlet edit -L \\
+                               -s /domain -t elem -n vcpu -v 4 \\
+			       -s /domain/vcpu -t attr -n placement -v static \\
+			       $vm_name.xml";
+        script_run "virsh undefine $vm_name" unless (script_run "virsh undefine --nvram $vm_name") == 0;
+        assert_script_run " ! virsh list --all | grep $vm_name";
+        assert_script_run "virsh define $vm_name.xml";
+        assert_script_run "virsh start $vm_name";
+    }
+
+    wait_guest_online($vm_name);
+    assert_script_run "ssh root\@$vm_name \"zypper in -y irqbalance\"" unless script_run("ssh root\@$vm_name \"rpm -q irqbalance\"") eq 0;
+
+}
+
+# return the total NIC IRQs on each CPU core from guest
+sub get_nic_irqs_distribution_from_guest {
+    my ($vm_name, $cpu_account) = @_;
+
+    my @total_irqs_on_cpu = ();
+    my $irq_output = script_output("ssh root\@$vm_name \"grep vif0 /proc/interrupts\"");
+    record_info("NIC IRQs distribution", $irq_output);
+    for (my $cpu_id = 0; $cpu_id < $cpu_account; $cpu_id++) {
+        my @irqs_on_one_cpu = split('\n', script_output("ssh root\@$vm_name \"awk '/vif0/{ print \\\$(($cpu_id+2)) }' /proc/interrupts\""));
+        # sum of IRQs on one cpu core
+        for (my $i = 0; $i <= $#irqs_on_one_cpu; $i++) {
+            $total_irqs_on_cpu[$cpu_id] += $irqs_on_one_cpu[$i];
+        }
+    }
+
+    return @total_irqs_on_cpu;
+}
```

Comment:
Does /proc/irq/default_smp_affinity  have any impact on your test

Reply:
I don't think so. the value of /proc/irq/default_smp_affinity varies in SLE releases, but the effect if IRQ balance is similiar. So I guess different balancing rules are used in SLE releases, and the default_smp_affinity is not same accordingly.

Reply:
Do you think it is beneficial to output value of default_smp_affinity as well in your test ?

Reply:
yes. agree.

Reply:
the content of the default_smp_affinity from all SLE release is 'f' in the case of 4 cpu

Reply:
done